src/Shipments/PriceCalculator.php

```diff
@@ -3,155 +3,134 @@
+            throw new InvalidResourceException(
+                'Cannot calculate shipment price without a valid shipment weight.'
+            );
+        }
+
+        if ($serviceRate === null) {
+            $serviceRate = $this->determineServiceRateForShipment($shipment);
         }
 
-        if ($contract === null) {
+        if ($shipment->getPhysicalProperties()->getWeight() < $serviceRate->getWeightMin()
+            || $shipment->getPhysicalProperties()->getWeight() > $serviceRate->getWeightMax()) {
             throw new CalculationException(
-                'Cannot calculate a price for given shipment without a contract'
+                'Could not calculate price for the given service rate since it does not support the shipment weight.'
             );
         }
 
-        return $this->calculateGroupPrice($shipment, $contract)
-            + $this->calculateOptionsPrice($shipment, $contract);
+        if (empty($shipment->getServiceOptions())) {
+            return $serviceRate->getPrice();
+        }
+
+        return $this->calculateOptionsPrice($shipment, $serviceRate) + $serviceRate->getPrice();
     }
 
     /**
-     * Calculate the price based on the weight group for given shipment.
-     * Optionally a contract can be supplied for the price calculations. If no
-     * contract is given, the contract on the shipment is used.
+     * Calculate the price based on the selected options for given shipment.
      *
-     * @param ShipmentInterface             $shipment
-     * @param ServiceContractInterface|null $contract
+     * @param ShipmentInterface         $shipment
+     * @param ServiceRateInterface|null $serviceRate
      * @return int
      */
-    public function calculateGroupPrice(ShipmentInterface $shipment, ServiceContractInterface $contract = null)
+    public function calculateOptionsPrice(ShipmentInterface $shipment, ServiceRateInterface $serviceRate = null)
     {
-        if ($contract === null) {
-            $contract = $shipment->getServiceContract();
+        if ($shipment->getPhysicalProperties() === null
+            || $shipment->getPhysicalProperties()->getWeight() === null
+            || $shipment->getPhysicalProperties()->getWeight() < 0) {
+            throw new InvalidResourceException(
+                'Cannot calculate shipment price without a valid shipment weight.'
+            );
         }
 
-        if ($contract === null) {
-            throw new CalculationException(
-                'Cannot calculate a price for given shipment without a contract'
-            );
+        if ($serviceRate === null) {
+            $serviceRate = $this->determineServiceRateForShipment($shipment);
         }
 
-        if ($shipment->getWeight() < 0) {
+        if ($shipment->getPhysicalProperties()->getWeight() < $serviceRate->getWeightMin()
+            || $shipment->getPhysicalProperties()->getWeight() > $serviceRate->getWeightMax()) {
```

Comment:
You don't actually need a weight to calculate the price for the service options, so I would remove this check here.

Reply:
I do, actually. You can call this method without a service rate as second parameter. We will then determine the service rate for you based on the `service`, `contract` and `weight` of your shipment.

Reply:
But then if you move the weight check to `validateShipments()` as described above, you don't need to check it here anymore.